### DISCUSSION TITLE
Recover all Teslamate data

### DISCUSSION THREAD

**Discussion by alezanott** [teslamate-org/teslamate#2289] [Q&A]
My Raspberry Pi 3b+ decided to not boot anymore. When trying the booting (connected via HDMI), I can clearly see the raspberry logo then it stucks on the classic black screen with a ticking `-`

I want to recover all the data from Teslamate, as I want to upgrade to a Rasp PI 4. Is there any way I can do this? Thank you in advance.

**Comment by martinh2011** [comment]
I hardly dare to ask: Do you have a current backup of the Teslamate database? If not, it‘s difficult, maybe impossible.

I‘ve never done this myself, so I‘m basically speculating here. docker usually stores all data in /var/lib/docker, the data volumes being in /var/lib/docker/volumes. To access /var/lib/docker without the raspi running you would need another (linux) system where you could mount the SD card. If you can do that and the folder is still accessible you might be able to copy it into another docker installation, e.g. on your new raspi 4. docker daemon must not be running on the target system and probably should be the same version as on the old one. 

However, the raspi not starting could be a dead SD card, so you might be out of luck anyway 🙁

**Reply by alezanott** [reply]
Nope, never done any backup of it :-(
I have found the docker folder in the partition but I don't have the permissions to access it. I have tried giving it the chmod 777 but still it didn't work. Actually I can see all the folders in the root partition of the sd. Tried dd all the sd content creating an img on my linux machine, then copying the img to a new sd but it didn't work.

**Reply by martinh2011** [reply]
I would have expected that you have access to everything as root. But if the SD failed and some data is missing, who knows what the effect will be.
 Maybe this helps? https://raspberrypi.stackexchange.com/questions/7489/recover-files-from-broken-sd-card-no-boot

**Reply by alezanott** [reply]
I have succeed in copying the docker folder. Now How can I set a new docker instance on the new rasp, using the old files? Thank you in advance.

I didn't succeed in copying some file in the docker/overlay2 folder and the file backingFsBlockDev in the docker/volumes, but actually I think that this will not be a problem. I have checked in the volumes folder and found the DB of Teslamate.

@adriankumpf can you help me?

**Reply by yeaaaaaahh** [reply]
Any update here @alezanott I'm in the same situation as you, managed to copy across docker/volumes without issue but the symlink files in /overlay2/*/diff/etc/ssl/certs/ prevented the copy of the entire directory.
And trying to start without that was not successful.

**Reply by alezanott** [reply]
> Any update here @alezanott I'm in the same situation as you, managed to copy across docker/volumes without issue but the symlink files in /overlay2/*/diff/etc/ssl/certs/ prevented the copy of the entire directory. And trying to start without that was not successful.

Hi there! No luck in recovering the container. I have created a new container from scratch. I have solved the rpi issue related to the crash of Teslamate due to the fact that the RAM gets saturated quite easely. By having an automatic reboot schedule every 3 days Teslamate works great.

**Comment by mnadvornik** [comment]
@alezanott @yeaaaaaahh I highly discourage copying your docker container from one system to another. The whole reason behind containers are that they are disposable. The very only thing you need to recover teslamate is the volume containing the postgresql database. In case you can still access the postgresql container on the old system the easiest method would be to call `pg_dump` inside the container. If you can't access the postgresql container and if your file system has not been damaged in this area, you should be fine to recover the postgres volume directory and just start the container with the old volume.
Should the postgresql data directory not be fully readable or you had a crash that corrupted the database itself, then you will need to look into the its documentation or probably seek help in forums dedicated postgresql. You could start looking here: https://stackoverflow.com/a/11040122

Hope this helps

**Reply by alezanott** [reply]
> @alezanott @yeaaaaaahh I highly discourage copying your docker container from one system to another. The whole reason behind containers are that they are disposable. The very only thing you need to recover teslamate is the volume containing the postgresql database. In case you can still access the postgresql container on the old system the easiest method would be to call `pg_dump` inside the container. If you can't access the postgresql container and if your file system has not been damaged in this area, you should be fine to recover the postgres volume directory and just start the container with the old volume. Should the postgresql data directory not be fully readable or you had a crash that corrupted the database itself, then you will need to look into the its documentation or probably seek help in forums dedicated postgresql. You could start looking here: https://stackoverflow.com/a/11040122
> 
> Hope this helps

Thank you for your reply! Unfortunately I didn't have had the possibility to do a `pg_dump`. At the end I decided to run from scratch a new container. I'm now transferring all the data from a rpi to a windows server, hoping to not lose nothing.
